crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -1872,10 +1872,12 @@ macro_rules! impl_anytuple_fetch {
                         $name::update_component_access($name, &mut intermediate);
                         _new_access.append_or(&intermediate);
                         _new_access.extend_access(&intermediate);
+                        _access.extend_access(&intermediate); // add accesses for remaining items to check against
                     } else {
                         $name::update_component_access($name, &mut _new_access);
                         _new_access.required = _access.required.clone();
                         _not_first = true;
+                        _access.extend_access(&_new_access); // add accesses for remaining items to check against
```

Comment:
The comment should make it more explicit why this works. At first glance it might appear that this simply works because it's adding the access to the output `_access`, but actually `_access` will be replaced unconditionally at the end of the function by `_new_access`. The reason this works instead is because `_access` is cloned in the line 1871 (unfortunately I couldn't select it...), and by adding the access to it the later calls to `update_component_access` will see it and will be able to panic if they detect conflicting access. However this is not clear at all! Moreover the alternating use of `_new_access`, `_access` and `intermediate`, while necessary, is pretty difficult to reason with. IMO this really needs some better comments explaining how it all works and why it's ok.

Reply:
I agree with the request for better comments. I also think this is overly constrained, but I'm very okay with restricting hypothetical uses for now to avoid UB.

Reply:
I agree that in general access-related comments are sorely lacking. Every time I look at access-related code it takes me a super long time to understand why the code was written this way (generally subtle issues where we want to update the `read/write` access but not the `with/without` filters)

For example in this code, I don't get the purpose of the 
`_new_access.required = _access.required.clone();` line.
Is it because `required` shouldn't be updated at all, since none of the `AnyOf` components should be required?